pkg/master/controller/crdregistration/crdregistration_controller.go

```diff
@@ -182,11 +184,12 @@ func (c *crdRegistrationController) processNextWorkItem() bool {
 }
 
 func (c *crdRegistrationController) enqueueCRD(crd *apiextensions.CustomResourceDefinition) {
-	c.queue.Add(schema.GroupVersion{Group: crd.Spec.Group, Version: crd.Spec.Version})
+	for _, version := range crd.Spec.Versions {
+		c.queue.Add(schema.GroupVersion{Group: crd.Spec.Group, Version: version.Name})
+	}
 }
 
 func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.GroupVersion) error {
-	found := false
 	apiServiceName := groupVersion.Version + "." + groupVersion.Group
```

Comment:
rework this... a single version and apiServiceName isn't sufficient with the potential for multiple versions

Reply:
hmm... it looks like the loop in `enqueueCRD` will ensure this is called for every present version... but I don't think that does the right thing on updates that remove served versions

Reply:
definitely need a test that goes through a version lifecycle to flush out issues, something like:

1. create with `v1`, ensure v1 appears in discovery, is usable, and persists an object as v1
2. add `v2` with served=true, ensure v1 and v2 appear in discovery, are usable, and still persist as v1
3. set `v2` stored=true, ensure new object persists as v2, update all existing objects and ensure they persist as v2
4. set v1 served=false, ensure v1 disappears from discovery and from api handler
5. remove v1 from status.storedVersions
6. remove v1 from versions

Reply:
A nice test to have, will add it.

Reply:
I hope I can do it after #63693 but will add this test in next round if that PR was not merged yet.

Reply:
after further thought, if we enqueue all versions in the old and new CRD on update, this will do the right thing